cln-rpc/src/primitives.rs

```diff
@@ -0,0 +1,132 @@
+use anyhow::{anyhow, Error, Result};
+use serde::Deserializer;
+use serde::{Deserialize, Serialize};
+#[derive(Copy, Clone, Serialize, Deserialize, Debug)]
+#[allow(non_camel_case_types)]
+pub enum ChannelState {
+    OPENINGD,
+    CHANNELD_AWAITING_LOCKIN,
+    CHANNELD_NORMAL,
+    CHANNELD_SHUTTING_DOWN,
+    CLOSINGD_SIGEXCHANGE,
+    CLOSINGD_COMPLETE,
+    AWAITING_UNILATERAL,
+    FUNDING_SPEND_SEEN,
+    ONCHAIN,
+    DUALOPEND_OPEN_INIT,
+    DUALOPEND_AWAITING_LOCKIN,
+}
+
+#[derive(Copy, Clone, Serialize, Deserialize, Debug)]
+#[allow(non_camel_case_types)]
+pub enum ChannelStateChangeCause {
+    UNKNOWN,
+    LOCAL,
+    USER,
+    REMOTE,
+    PROTOCOL,
+    ONCHAIN,
+}
+
+#[derive(Copy, Clone, Serialize, Debug, PartialEq)]
+pub enum Amount {
+    Millisatoshi(u64),
+    Satoshi(u64),
+    Millibitcoin(u64),
+    Bitcoin(u64),
+}
```

Comment:
That type seems cursed, why not just go with some msat-based amount type and convert appropriately? Is re-encoding with syntactical (instead of just semantic) equality really a goal?

Reply:
We've had quite some cases in the past where users of the API misunderstood the unit they were dealing with, and ended up creating channels that were 1000x the size they intended, or transferred 0.001x the amount they were supposed to. Hence this more complex, but also much more strict and user visible type.

Reply:
Do you think it's that much better than e.g. `struct Amount { pub msat: u64 }` with `from_btc(btc: u64)`, `from_sats(sats: u64)` and `from_milli_sats(msats: u64)` as constructors?

Its one thing to misjudge units in the CLI when not supplying units at all, it's a completely different thing working with an API that forces you to specify your intent through either function names or variants, imo both work pretty much the same, this will just be somewhat unergonomic imo, imagine maybe implementing arithmetic on this type :grimacing:

At least there should be a `to_msat(&self) -> u64` function so users don't need to match every time.

Reply:
The main reason we have this rather complex type is because we want to have `Amount::All` and `Amount::Any` variants as well, which we can't map to a `pub struct Amount { millisatoshi: u64, }`

Reply:
Maybe it would still make sense to do

```rust
enum Amount {
    MilliSat(u64),
    All,
    Any
}
```

Btw, does `All` even exist on the LN side? I only know it from the on-chain wallet. If so, maybe even split into two:

```rust
enum BtcAmount {
    Amount(bitcoin::Amount), // only supports sat granularity
    All
}

enum LnAmount {
    MilliSat(u64),
    Any
}
```

Otherwise there will be weird error conditions depending on context.

Reply:
Do you think that dividing the enum in ln and btc makes the API harder to read? I mean, there is any ln amount, just a milli convention. 

I like the @cdecker idea, to have one single wrapper around the amount value, at the end of the day we want that the user can use the library with any knowledge of btc.

Reply:
@vincenzopalazzo I think encoding invariants in the types makes the API easier to use. If you were to supply an msat granular amount to an operation expecting sat granularity that would need to be an error or panic (both annoying imo). Same goes for the possible `all` and `any` variants.

@cdecker reading the `lightning-cli` docs I noticed that other parameters also take `all` (e.g. the id when setting channel fees). So another idea would be to define the amount type as msat without regard to `all`/`any`, but also defining a `MaybeAll` and `MaybeAny` enum with a custom `serde` impl.

```rust
enum MaybeAll<T> {
    Concrete(T),
    All
}
```

Reply:
I very much like the idea of using `MaybeAll<T>` and `MaybeAny<T>`, however currently it is not possible for us to use these, since this is information is currently not conveyed in the JSON schema at all (flat `msat` type for all fields dealing with amounts).

To fix this we'd have to go through the RPC calls and annotate them with `msat-any` or `msat-all` if one of these variants is to be used. So before going that way it's good to be sure we want that option.

Reply:
After looking a bit harder at the JSON schema, I think we can annotate the very few cases where we'd have `msat|all` and `msat|any`. My count shows that only 3 RPC methods accept `any` and 5 accept `all`, and since I need to write the request schemata anyway, that might be a good time to introduce these. Trying the changes in a separate branch and they look rather nice.

The only thing that is less nice is that we don't remember the unit that the user chose to express in, flattening everything to msats, but that's a presentation issue anyway.

Reply:
> The only thing that is less nice is that we don't remember the unit that the user chose to express in, flattening everything to msats, but that's a presentation issue anyway.

Imo representation should be left to the application (there's probably an application-wide standard anyway, so no need to preserve units used). On an API level having only msat makes a lot of sense because it reduces complexity and computers don't care. I assume having support for different units in the first place was to make life easier for `lightning-cli` users. Thinking of it, was any thought given to impl the unit conversion in `lightning-cli` instead of `lightningd` (or whatever subdaemon it is)?

Reply:
Agreed, I think the `struct Amount { msat: u64 }` is likely far better than what we have so far. I'll test that one and see how much fallout it produces.